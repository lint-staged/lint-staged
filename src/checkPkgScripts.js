'use strict'

const chalk = require('chalk')
const dedent = require('dedent')
const has = require('lodash/has')

const warn = msg => {
  console.warn(chalk.yellowBright.bold(msg))
}

/**
 * Checks if the given command or binary name is present in the package.json scripts. This would be
 * called if and when resolving a binary fails in `findBin`.
 *
 * @param {Object} pkg package.json
 * @param {string} cmd
 * @param {string} binName
 * @param {Array<string>} args
 * @throws {Error} If a script is found in the pkg for the given `cmd` or `binName`.
 */
module.exports = function checkPkgScripts(pkg, cmd, binName, args) {
  if (pkg && pkg.scripts) {
    const { scripts } = pkg
    let scriptName
    let script
    if (has(scripts, cmd)) {
      scriptName = cmd
      script = scripts[cmd]
    } else if (has(scripts, binName)) {
      scriptName = binName
      script = scripts[binName]
    } else {
      return
    }

    const argsStr = args && args.length ? args.join(' ') : ''
    warn(dedent`
      \`lint-staged\` no longer supports running scripts defined in package.json.

      The same behavior can be achieved by changing the command to any of the following:
        - \`npm run ${scriptName} -- ${argsStr}\`
        - \`${script} ${argsStr}\`
    `)
    throw new Error(`Could not resolve binary for \`${cmd}\``)
  }
}
