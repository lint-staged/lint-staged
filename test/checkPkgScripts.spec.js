import makeConsoleMock from 'consolemock'
import checkPkgScripts from '../src/checkPkgScripts'

describe('checkPkgScripts', () => {
  describe('does not throw', () => {
    it('if there are no scripts defined', () => {
      expect(() => checkPkgScripts(null)).not.toThrow()
      expect(() => checkPkgScripts({})).not.toThrow()
    })

    it('if the script cannot be found', () => {
      expect(() => checkPkgScripts({ scripts: {} }, 'fmt')).not.toThrow()
    })
  })

  describe('throws', () => {
    const originalConsole = global.console
    beforeAll(() => {
      global.console = makeConsoleMock()
    })

    beforeEach(() => {
      global.console.clearHistory()
    })

    afterAll(() => {
      global.console = originalConsole
    })

    const pkg = {
      scripts: {
        lint: 'eslint .'
      }
    }

    it('if the cmd is defined in script', () => {
      expect(() => {
        checkPkgScripts(pkg, 'lint', 'lint')
      }).toThrow('Could not resolve binary for `lint`')
      expect(console.printHistory()).toMatchSnapshot()
    })

    it('if the binary name is defined in script', () => {
      expect(() => {
        checkPkgScripts(pkg, 'lint --fix', 'lint', ['--fix'])
      }).toThrow('Could not resolve binary for `lint --fix`')
      expect(console.printHistory()).toMatchSnapshot()
    })
  })
})
