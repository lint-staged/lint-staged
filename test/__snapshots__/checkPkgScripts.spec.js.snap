// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkPkgScripts throws if the binary name is defined in script 1`] = `
"
WARN \`lint-staged\` no longer supports running scripts defined in package.json.

The same behavior can be achieved by changing the command to any of the following:
  - \`npm run lint -- --fix\`
  - \`eslint . --fix\`"
`;

exports[`checkPkgScripts throws if the cmd is defined in script 1`] = `
"
WARN \`lint-staged\` no longer supports running scripts defined in package.json.

The same behavior can be achieved by changing the command to any of the following:
  - \`npm run lint -- \`
  - \`eslint . \`"
`;
